core/internal/filetransfer/file_transfer_default.go

```diff
@@ -77,10 +77,20 @@ func (ft *DefaultFileTransfer) Upload(task *Task) error {
 		)
 	}
 
-	task.Size = stat.Size()
+	if task.Offset+task.Size > stat.Size() {
+		// If the range exceeds the file size, there was some kind of error upstream.
+		return fmt.Errorf("file transfer: upload: offset + size exceeds the file size")
+	}
+
+	if task.Offset == 0 && task.Size == 0 {
+		// If both offset and size are zero (default), upload the entire file.
+		task.Size = stat.Size()
+	}
```

Comment:
I think the semantics would be simpler if the special case were `task.Size == 0` alone, rather than the combination of `task.Offset == 0` and `task.Size == 0`. So

```suggestion
	if task.Size == 0 {
		task.Size = stat.Size() - task.Offset
	}
```

(this also matches the docs)